rbac/management/role/view.py

```diff
@@ -440,28 +453,71 @@ def update(self, request, *args, **kwargs):
         validate_uuid(kwargs.get("uuid"), "role uuid validation")
         self.validate_role(request)
 
-        update_role = self.update_with_relation_api_replication(request=request, args=args, kwargs=kwargs)
-
-        if status.is_success(update_role.status_code):
-            auditlog = AuditLog()
-            role = self.get_object()
-            auditlog.log_edit(request, AuditLog.ROLE, role)
-
-        return update_role
-
-    def update_with_relation_api_replication(self, request, *args, **kwargs):
-        """Update a role with replicating data into Relation API."""
         try:
-            role = self.get_object()
-            dual_write_handler = RelationApiDualWriteHandler(role, "UPDATE")
-            dual_write_handler.generate_relations_from_current_state_of_role()
             with transaction.atomic():
-                response = super().update(request=request, args=args, kwargs=kwargs)
-                dual_write_handler.generate_replication_event_to_outbox(self.get_object())
+                return super().update(request=request, args=args, kwargs=kwargs)
         except DualWriteException as e:
             return self.dual_write_exception_response(e)
 
-        return response
+    def perform_create(self, serializer):
+        """
+        Create the role and publish outbox, notification, and audit events.
+
+        Assumes concurrent updates are prevented (e.g. with atomic block and locks).
+        """
+        role = serializer.save()
+
+        dual_write_handler = RelationApiDualWriteHandler(role, "CREATE")
+        dual_write_handler.generate_replication_event_to_outbox(role)
+
+        role_obj_change_notification_handler(role, "created", self.request.user)
+
+        auditlog = AuditLog()
+        auditlog.log_create(self.request, AuditLog.ROLE)
+
+    def perform_update(self, serializer):
+        """
+        Update the role and publish outbox, notification, and audit events.
+
+        Assumes concurrent updates are prevented (e.g. with atomic block and locks).
+        """
+        if self.action != "partial_update":
+            dual_write_handler = RelationApiDualWriteHandler(serializer.instance, "UPDATE")
+            dual_write_handler.load_relations_from_current_state_of_role()
+
+        role = serializer.save()
```

Comment:
One thing to note about this happening before serializer.save() here is that this load will happen before some validation in the serializer happens. The serializer has a `validate_role_update` method which checks if the role is a system role. This could fail and we could have saved ourselves from building up current mappings.

It's tempting to move that validation here, which would make it more cohesive with the same validation that happens in `perform_destroy` but I'll leave that up to others reviewing.